tokio/src/stream/fuse.rs

```diff
@@ -0,0 +1,54 @@
+use crate::stream::Stream;
+
+use pin_project_lite::pin_project;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+pin_project! {
+    /// Stream returned by [`fuse()`][super::StreamExt::fuse].
+    #[derive(Debug)]
+    pub struct Fuse<T> {
+        #[pin]
+        stream: Option<T>,
+    }
+}
+
+impl<T> Fuse<T>
+where
+    T: Stream,
+{
+    pub(crate) fn new(stream: T) -> Fuse<T> {
+        Fuse { stream: Some(stream) }
+    }
+}
+
+impl<T> Stream for Fuse<T>
+where
+    T: Stream,
+{
+    type Item = T::Item;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<T::Item>> {
+        let res = unsafe {
```

Comment:
im assuming pin project doesn't support pinning in the option?

Reply:
nope.

Reply:
can use option::as_pin_mut instead: https://doc.rust-lang.org/nightly/core/option/enum.Option.html#method.as_pin_mut

Reply:
Well that is great to know :+1: